File: B_PL_Codes/16-CódigoComFunções/6-SerieDadosAnalise/Rotinas3SerieDados.c


int CarregarPares(int numeros[], int pares[], int tamanho)
{
    // Rotina usada para preencher um vetor pre-definido no programa principal
    int cont = 0;
    for (int i = 0; i < tamanho; i++)
    {
        if (numeros[i] % 2 == 0)
        {
            pares[cont] = numeros[i];
            cont++;
        }
    }
    return cont;
}

int CarregarPrimos(int numeros[], int primos[], int tamanho)
{
    // Rotina usada para preencher um vetor pre-definido no programa principal
    int cont = 0;
    for (int i = 0; i < tamanho; i++)
    {
        int primo = 1;
        if (numeros[i] <= 2)
        {
            primo = 0;
        }
        for (int j = 2; j <= numeros[i] / 2; j++)
        {
            if (numeros[i] % j == 0)
            {
                primo = 0;
                break;
            }
        }
        if (primo)
        {
            primos[cont] = numeros[i];
            cont++;
        }
    }
    return cont;
}

File: B_PL_Codes/16-CódigoComFunções/6-SerieDadosAnalise/inJava/Rotinas3SerieDados.java

public class Rotinas3SerieDados {

    static int[] CarregarPares(int numeros[]) {
        // 1. contar pares
        int cont = 0;
        for (int i = 0; i < numeros.length; i++) {
            if (numeros[i] % 2 == 0) {
                cont++;
            }
        }
        // 2. criar vetor com tamanho exato
        int[] pares = new int[cont];

        // 3. preencher vetor
        cont = 0;
        for (int i = 0; i < numeros.length; i++) {
            if (numeros[i] % 2 == 0) {
                pares[cont] = numeros[i];
                cont++;
            }
        }
        return pares;

    }

    static int[] CarregarPrimos(int[] numeros) {
        // Estratégia:
        // 1) contar os primos; 2) Criar um vetor de primos; 3) Carregar o vetor;

        // 1ª etapa — contar primos
        int cont = 0;
        for (int i = 0; i < numeros.length; i++) {
            boolean primo = true;
            if (numeros[i] < 2) {
                primo = false;
            }
            for (int j = 2; j <= numeros[i] / 2 && primo; j++) {
                if (numeros[i] % j == 0) {
                    primo = false;
                }
            }
            if (primo) {
                cont++;
            }
        }
        // 2ª etapa — alocar vetor com tamanho exato
        int[] primos = new int[cont];

        // 3ª etapa — preencher vetor
        cont = 0;
        for (int i = 0; i < numeros.length; i++) {
            boolean primo = true;
            if (numeros[i] < 2) {
                primo = false;
            }
            for (int j = 2; j <= numeros[i] / 2 && primo; j++) {
                if (numeros[i] % j == 0) {
                    primo = false;
                }
            }
            if (primo) {
                primos[cont] = numeros[i];
                cont++;
            }
        }
        return primos;
    }

}